File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        maxD = 0

        def dfs(n: Optional[TreeNode]):
            nonlocal maxD

            if n is None:
                return

            dfs(n.left)
            dfs(n.right)

            d = 0

            if n.left:
                d += 1 + height(n.left)
                
            if n.right:
                d += 1 + height(n.right)
            
            maxD = max(maxD, d)


        def height(n: Optional[TreeNode]) -> int:
            if n is None or (n.left is None and n.right is None):
                return 0
            
            return max(height(n.left), height(n.right)) + 1

        dfs(root)

        return maxD

File: Data Structures & Algorithms/trapping-rain-water/submission-2.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        
        int w = 0;

        auto l = height.begin(), r = height.end() - 1;

        int mh = min(*l, *r);

        while (l < r) {
            mh = max(mh, min(*l, *r));

            if (*l < *r) {
                w += max(0, mh - *l);
                l++;
            } else {
                w += max(0, mh - *r);
                r--;
            }
        }

        return w;
    }
};
